### DISCUSSION TITLE
How to use fbinding and ibinding in the lfe_env?

### DISCUSSION THREAD

**Discussion by lin7sh** [lfe/lfe#483] [Q&A] [answered]
I just figure out how to use vbinding and mbinding, but not the fbinding
```
lfe> (set env (lfe_env:add_fbinding 'add 1 '(lambda (x) (+ x 1)) (lfe_env:new))) 
#(env #M() #M(add #(function (#(1 (lambda (x) (+ x 1)))))))
```
but I can't use it in eval
```
lfe> (eval '(add 1) env)
exception error: no function clause matching

```
I think it might be caused by dynamic_expr which I can't find any information either, and what is ibinding?

**Comment by rvirding** [comment, marked as answer]
The issue here is that the `lfe_env` module is really an internal module and shouldn't be used directly used. Check in `lfe_macro.erl` to see how `defun` is expanded and in `lfe_shell.erl` how `define-function` is implemented.
